updates to ray_aws config keys, example apps and doc

jieru-hu opened the pull request:
<!-- Thank you for sending a PR and taking the time to improve Hydra -->

## Motivation
updates to the example app so least changes needed to run the example app in 3 modes.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebookresearch/hydra/blob/master/CONTRIBUTING.md)?

Yes

## Test Plan

run the example apps :) 

## Related Issues and PRs

(Is this PR part of a group of changes? Link the other relevant PRs and Issues here. Use https://help.github.com/en/articles/closing-issues-using-keywords for help on GitHub syntax)

omry commented on website/docs/plugins/ray_launcher.md:
````diff
@@ -223,7 +223,7 @@ hydra.launcher.ray_cluster_cfg.provider.cache_stopped_nodes=true
  The [example application](https://github.com/facebookresearch/hydra/tree/master/plugins/hydra_ray_launcher/examples/simple) starts a new ray cluster. 
 ```commandline
-$ python my_app.py  --multirun hydra/launcher=ray
+$ python simple/my_app.py  --multirun hydra/launcher=ray
````
Why not run it from within simple?

omry commented on plugins/hydra_ray_launcher/examples/simple/config.yaml:
```diff
@@ -3,7 +3,7 @@ defaults:
   launcher:
-    ray_remote_cfg:
-      num_cpus: 1
+    ray_init_cfg:
+      address: null
```
I meant adding this to the plugin's config. something like:
```python
@dataclass
class RayLauncherConf:
    _target_: str = "hydra_plugins.hydra_ray_launcher.ray_launcher.RayLauncher"
    ray_init_cfg: Dict[str, Any] = field(default_factory={"address": "null"})
    ray_remote_cfg: Dict[str, Any] = field(default_factory=dict)
```

jieru-hu commented:
This PRs addresses a few things:
1. shorten the ray_aws config keys, in particular
```
s/ray_init_cfg/ray_init/g
s/ray_cluster_cfg/ray_cluster/g
s/ray_remote_cfg/ray_remote/g
```
2. Add ray_init default address
3. minor updates to the example app
4. updates to documentation.

omry commented on plugins/hydra_ray_launcher/hydra_plugins/hydra_ray_launcher/_launcher_util.py:
```diff
@@ -46,13 +46,13 @@ def _run_job(
-        run_job_ray = ray.remote(**ray_remote_cfg)(_run_job)
+    if ray_remote:
+        run_job_ray = ray.remote(**ray_remote)(_run_job)
     else:
```
seeing this makes me wonder if the config should look like:
```yaml
hydra:
  launcher:
    ray:
      init:
        ..
      remote:
        ..
 ```

This seems to model things better. and if we ever want to configure another Ray function we won't need to repeat the ray_ prefix again everywhere.

jieru-hu replied:
that's a great idea!

I'm actually working on adding a new node (`run_env`) which would also need the prefix probably if we go with the current way. 

will modify the configs.

omry replied:
awesome :)

omry commented on plugins/hydra_ray_launcher/hydra_plugins/hydra_ray_launcher/conf/__init__.py:
```diff
@@ -15,8 +15,8 @@
     _target_: str = "hydra_plugins.hydra_ray_launcher.ray_launcher.RayLauncher"
-    ray_init_cfg: Dict[str, Any] = field(default_factory=dict)
-    ray_remote_cfg: Dict[str, Any] = field(default_factory=dict)
+    ray_init: Dict[str, Any] = field(default_factory=lambda: {"address": None})
```
to make this easier to read, add a trailing comma:
```
ray_init: Dict[str, Any] = field(default_factory=lambda: {"address": None,})
```
and let black format it.

jieru-hu replied:
nice!

omry commented on plugins/hydra_ray_launcher/hydra_plugins/hydra_ray_launcher/conf/__init__.py:
```diff
@@ -188,9 +188,9 @@ class RayAWSLauncherConf:
-    ray_init_cfg: Dict[str, Any] = field(default_factory=dict)
-    ray_remote_cfg: Dict[str, Any] = field(default_factory=dict)
-    ray_cluster_cfg: RayClusterConf = RayClusterConf()
+    ray_init: Dict[str, Any] = field(default_factory=lambda: {"address": None})
+    ray_remote: Dict[str, Any] = field(default_factory=dict)
+    ray_cluster: RayClusterConf = RayClusterConf()
```
If we went with my proposal to have a ray node, this would be:
```python
@dataclass
class RayConfig:
    init: Dict[str, Any] = field(default_factory=lambda: {"address": None})
    ray_remote: Dict[str, Any] = field(default_factory=dict)
    cluster: RayClusterConf = RayClusterConf()

@dataclass
class RayLauncherConf:
  ray : RayConfig = RayConfig()
  ..

@dataclass
RayAWSLauncherConf
  ray : RayConfig = RayConfig()
  ..
```

jieru-hu commented:
Changed the config to look like the following, in particular extra stuff launcher provides falls under `hydra.launcher`; configuration passed into  ray directly grouped under  `hydra.launcher.ray`

```
# @package hydra.launcher
_target_: hydra_plugins.hydra_ray_launcher.ray_aws_launcher.RayAWSLauncher
mandatory_install:
  hydra_version: 1.0.3
  ray_version: 1.0.1.post1
  cloudpickle_version: 1.6.0
  omegaconf_version: 2.1.0dev13
  pickle5_version: 0.0.11
  hydra_ray_launcher_version: 0.1.2
  install_commands:
  - conda create -n hydra_${python_version:micro} python=${python_version:micro} -y
  - echo 'export PATH="$HOME/anaconda3/envs/hydra_${python_version:micro}/bin:$PATH"'
    >> ~/.bashrc
  - pip install omegaconf==${hydra.launcher.mandatory_install.omegaconf_version}
  - pip install hydra-core==${hydra.launcher.mandatory_install.hydra_version}
  - pip install https://hydra-test-us-west-2.s3-us-west-2.amazonaws.com/hydra_ray_launcher-0.1.2-py3-none-any.whl
  - pip install ray==${hydra.launcher.mandatory_install.ray_version}
  - pip install cloudpickle==${hydra.launcher.mandatory_install.cloudpickle_version}
  - pip install pickle5==${hydra.launcher.mandatory_install.pickle5_version}
ray:
  init:
    address: null
  remote: {}
  cluster:
    cluster_name: default
    min_workers: 0
    max_workers: 1
    initial_workers: 0
    autoscaling_mode: default
    target_utilization_fraction: 0.8
    idle_timeout_minutes: 5
    docker:
      image: ''
      container_name: ''
      pull_before_run: true
      run_options: []
    provider:
      type: aws
      region: us-west-2
      availability_zone: us-west-2a,us-west-2b
      cache_stopped_nodes: false
      key_pair:
        key_name: hydra
    auth:
      ssh_user: ubuntu
    head_node:
      InstanceType: m5.large
      ImageId: ami-008d8ed4bd7dc2485
    worker_nodes:
      InstanceType: m5.large
      ImageId: ami-008d8ed4bd7dc2485
    file_mounts: {}
    initialization_commands: []
    setup_commands: []
    head_setup_commands: []
    worker_setup_commands: []
    head_start_ray_commands:
    - ray stop
    - ulimit -n 65536; ray start --head --redis-port=6379 --object-manager-port=8076
      --autoscaling-config=~/ray_bootstrap_config.yaml
    worker_start_ray_commands:
    - ray stop
    - ulimit -n 65536; ray start --address=$RAY_HEAD_IP:6379 --object-manager-port=8076
stop_cluster: true
sync_up:
  source_dir: null
  target_dir: null
  include: []
  exclude: []
sync_down:
  source_dir: null
  target_dir: null
  include: []
  exclude: []
```

omry commented:
> Changed the config to look like the following, in particular extra stuff launcher provides falls under `hydra.launcher`; configuration passed into ray directly grouped under `hydra.launcher.ray`
> 
> ```
> # @package hydra.launcher
> _target_: hydra_plugins.hydra_ray_launcher.ray_aws_launcher.RayAWSLauncher
> mandatory_install:
>   hydra_version: 1.0.3
>   omegaconf_version: 2.1.0dev13

Hydra 1.0.3 and OmegaConf 2.1 is not a supported combination.
Also, why 1.0.3 and not 1.0.4? (Or really, why not latest 1.0?)
For OmegaConf, you should install the version required by Hydra. 1.0 will install the latest OmegaConf 2.0.

omry commented on plugins/hydra_ray_launcher/hydra_plugins/hydra_ray_launcher/_launcher_util.py:
```diff
@@ -18,11 +18,11 @@
-        if ray_init_cfg:
-            ray.init(**ray_init_cfg)
+        log.info(f"Initializing ray with config: {init_cfg}")
+        if ray.init:
```
what is ray.init?

omry commented on plugins/hydra_ray_launcher/hydra_plugins/hydra_ray_launcher/ray_aws_launcher.py:
```diff
@@ -20,17 +17,13 @@ class RayAWSLauncher(Launcher):
-        self.ray_remote_cfg = ray_remote_cfg
-        self.ray_cluster_cfg = ray_cluster_cfg
-        self.docker_enabled = self.ray_cluster_cfg.docker.image != ""
+        self.ray = ray
```
ray sounds like the the ray module. can you assign to ray_cfg to make it clearer?

omry commented on website/docs/plugins/ray_launcher.md:
```diff
@@ -49,66 +49,65 @@ $ python your_app.py hydra/launcher=ray_aws --cfg hydra -p hydra.launcher
   hydra_version: 1.0.3
-  ray_version: 1.0.0
+  ray_version: 1.0.1.post1
   cloudpickle_version: 1.6.0
-  omegaconf_version: 2.1.0dev9
+  omegaconf_version: 2.1.0dev13
```
How is this interpreted?
you are appending _version to the package name. is Ray parsing this?

jieru-hu commented:
>Hydra 1.0.3 and OmegaConf 2.1 is not a supported combination.
Also, why 1.0.3 and not 1.0.4? (Or really, why not latest 1.0?)
For OmegaConf, you should install the version required by Hydra. 1.0 will install the latest OmegaConf 2.0.

My bad, this is my weird conda testing env. Users who follow the installation instruction would never run into this. fixed the example (running from a clean conda env)

>what is ray.init?

oops.. fixed. thanks for catching this


>ray sounds like the the ray module. can you assign to ray_cfg to make it clearer?

fixed! yea, it's a bad naming

>How is this interpreted?
you are appending _version to the package name. is Ray parsing this?

[here](https://github.com/facebookresearch/hydra/blob/master/plugins/hydra_ray_launcher/hydra_plugins/hydra_ray_launcher/conf/__init__.py#L83-L90) is where we get the versions.
We interprete these and dump them into the tmp yaml file we use to spin up ray cluster, ray will read the values and do the installation.

omry commented on plugins/hydra_ray_launcher/hydra_plugins/hydra_ray_launcher/conf/__init__.py:
```diff
@@ -12,11 +12,16 @@
 
+@dataclass
+class RayConf:
+    init: Dict[str, Any] = field(default_factory=lambda: {"address": None})
```
I know we have been through it but asking one last time:
can this dict be a Structured Config?

```
@dataclass
class RayInitConf:
  address: Optional[str] = None
```

jieru-hu replied:
let me address this as a follow up (ray init takes in a bunch of params)
I want to get this in soon so we can get the post unblocked :)

omry commented on plugins/hydra_ray_launcher/hydra_plugins/hydra_ray_launcher/ray_aws_launcher.py:
```diff
@@ -20,17 +17,13 @@ class RayAWSLauncher(Launcher):
-        self.ray_cluster_cfg = ray_cluster_cfg
-        self.docker_enabled = self.ray_cluster_cfg.docker.image != ""
+        self.ray_cfg = ray
+        self.docker_enabled = self.ray_cfg.cluster.docker.image != ""
```
None docker.image counts as enabled?

jieru-hu replied:
that's some ugly python code :)
fixed.

omry commented:
> > Hydra 1.0.3 and OmegaConf 2.1 is not a supported combination.
> > Also, why 1.0.3 and not 1.0.4? (Or really, why not latest 1.0?)
> > For OmegaConf, you should install the version required by Hydra. 1.0 will install the latest OmegaConf 2.0.
> 
> My bad, this is my weird conda testing env. Users who follow the installation instruction would never run into this. fixed the example (running from a clean conda env)
> 
> > what is ray.init?
> 
> oops.. fixed. thanks for catching this

Tests didn't catch it?

> > How is this interpreted?
> > you are appending _version to the package name. is Ray parsing this?
> 
> [here](https://github.com/facebookresearch/hydra/blob/master/plugins/hydra_ray_launcher/hydra_plugins/hydra_ray_launcher/conf/__init__.py#L83-L90) is where we get the versions.
> We interprete these and dump them into the tmp yaml file we use to spin up ray cluster, ray will read the values and do the installation.

So, how is this being used?
your question does not answer my question really. What are we doing with this input list?

Can you explain the flow?
the syntax is odd, why not match the pip requirement format?
```
hydra_core==1.0.4
```

jieru-hu commented on plugins/hydra_ray_launcher/hydra_plugins/hydra_ray_launcher/_launcher_util.py:
```diff
@@ -46,13 +46,13 @@ def _run_job(
 ) -> Any:
-    if ray_remote_cfg:
-        run_job_ray = ray.remote(**ray_remote_cfg)(_run_job)
+    if ray_remote:
```
yes this is checking to see if the DictConfig is empty; I think this is the standard syntax, should I be instead checking the len(ray_remote) > 0?

jieru-hu commented:
> > > Hydra 1.0.3 and OmegaConf 2.1 is not a supported combination.
> > > Also, why 1.0.3 and not 1.0.4? (Or really, why not latest 1.0?)
> > > For OmegaConf, you should install the version required by Hydra. 1.0 will install the latest OmegaConf 2.0.
> > 
> > 
> > My bad, this is my weird conda testing env. Users who follow the installation instruction would never run into this. fixed the example (running from a clean conda env)
> > > what is ray.init?
> > 
> > 
> > oops.. fixed. thanks for catching this
> 
> Tests didn't catch it?

`ray.init` is a valid python object, function. so unfortunately not in this case :( 


> 
> > > How is this interpreted?
> > > you are appending _version to the package name. is Ray parsing this?
> > 
> > 
> > [here](https://github.com/facebookresearch/hydra/blob/master/plugins/hydra_ray_launcher/hydra_plugins/hydra_ray_launcher/conf/__init__.py#L83-L90) is where we get the versions.
> > We interprete these and dump them into the tmp yaml file we use to spin up ray cluster, ray will read the values and do the installation.
> 
> So, how is this being used?
> your question does not answer my question really. What are we doing with this input list?
> 
> Can you explain the flow?
> the syntax is odd, why not match the pip requirement format?
> 
> ```
> hydra_core==1.0.4
> ```

Here is where the input list are used: https://github.com/facebookresearch/hydra/blob/master/plugins/hydra_ray_launcher/hydra_plugins/hydra_ray_launcher/conf/__init__.py#L92-L103

omry commented:
> ray.init is a valid python object, function. so unfortunately not in this case :(

This means that one of the two branch cases is not covered by tests, right?

jieru-hu commented:
> > ray.init is a valid python object, function. so unfortunately not in this case :(
> 
> This means that one of the two branch cases is not covered by tests, right?

yea :( cleaning up and adding my tests is on my todo list

jieru-hu commented:
PR updated per offline discussion

omry commented on website/docs/plugins/ray_launcher.md:
```diff
@@ -48,64 +48,61 @@ $ python my_app.py hydra/launcher=ray_aws --cfg hydra -p hydra.launcher
 # @package hydra.launcher
 _target_: hydra_plugins.hydra_ray_launcher.ray_aws_launcher.RayAWSLauncher
 mandatory_install:
```
mandatory_install is not a great name.
maybe `node_setup` or even just `setup`?

omry commented on plugins/hydra_ray_launcher/hydra_plugins/hydra_ray_launcher/conf/__init__.py:
```diff
@@ -84,26 +84,24 @@ class RsyncConf:
-    ).version
+class MandatoryInstallConf:
+    pip_packages: Dict[str, str] = field(
+        default_factory=lambda: {
+            "hydra_core": hydra.__version__,
+            "ray": ray.__version__,
+            "cloudpickle": cloudpickle.__version__,
+            "omegaconf": omegaconf.__version__,
+            "pickle5": pkg_resources.get_distribution("pickle5").version,
+            "hydra_ray_launcher": pkg_resources.get_distribution(
+                "hydra_ray_launcher"
+            ).version,
```
Ok, now I understand where the specific versions are coming from.
You can remove OmegaConf, installing Hydra will pull the proper version in.

jieru-hu replied:
OmegaConf is a key lib, i think it's worth it to pin the version so it's less likely to cause issue in the remote environment. (it might also be hard to debug when a version mismatch happens)

omry commented on plugins/hydra_ray_launcher/hydra_plugins/hydra_ray_launcher/conf/__init__.py:
```diff
@@ -84,26 +84,24 @@ class RsyncConf:
-    hydra_ray_launcher_version: str = pkg_resources.get_distribution(
-        "hydra_ray_launcher"
-    ).version
+class MandatoryInstallConf:
```
This class looks like EnvSetupConf more than it looks like MandatoryInstallConf :)

omry commented on plugins/hydra_ray_launcher/hydra_plugins/hydra_ray_launcher/conf/__init__.py:
```diff
@@ -197,7 +195,7 @@ class RayAWSConf(RayConf):
     _target_: str = "hydra_plugins.hydra_ray_launcher.ray_aws_launcher.RayAWSLauncher"
 
-    mandatory_install: PluginMandatoryInstallConf = PluginMandatoryInstallConf()
+    mandatory_install: MandatoryInstallConf = MandatoryInstallConf()
```
If you rename to EnvSetupConf:
```
env_setup : EnvSetupConf = EnvSetupConf()
```

omry commented on plugins/hydra_ray_launcher/hydra_plugins/hydra_ray_launcher/ray_aws_launcher.py:
```diff
@@ -20,17 +17,13 @@ class RayAWSLauncher(Launcher):
-        self.ray_cluster_cfg = ray_cluster_cfg
-        self.docker_enabled = self.ray_cluster_cfg.docker.image != ""
+        self.ray_cfg = ray
+        self.docker_enabled = True if self.ray_cfg.cluster.docker.image else False
```
will this work?
```suggestion
        self.docker_enabled = bool(self.ray_cfg.cluster.docker.image)
```

jieru-hu commented:
>mandatory_install is not a great name.
maybe node_setup or even just setup?

`hydra.ray.cluster` also has it's own `setup_commands` node
so how about something like the following
```yaml
hydra:
  launcher:
      env_setup: 
           pip_packages: {}
           commands: []
      ray:
         cluster:
            setup_commands: []
```

thoughts? @omry

omry commented:
> > mandatory_install is not a great name.
> > maybe node_setup or even just setup?
> 
> `hydra.ray.cluster` also has it's own `setup_commands` node
> so how about something like the following
> 
> ```yaml
> hydra:
>   launcher:
>       env_setup: 
>            pip_packages: {}
>            commands: []
>       ray:
>          cluster:
>             setup_commands: []
> ```
> 
> thoughts? @omry

Looks ok.

omry commented on plugins/hydra_ray_launcher/tests/test_ray_aws_launcher.py:
```diff
@@ -259,15 +259,15 @@ def manage_cluster() -> Generator[None, None, None]:
-                f"+hydra.launcher.ray_cluster_cfg.worker_nodes={ray_nodes_conf_override}",
-                f"+hydra.launcher.ray_cluster_cfg.head_node={ray_nodes_conf_override}",
+                f"hydra.launcher.ray.cluster.provider.key_pair.key_name=hydra_test_{cluster_name}",
+                "hydra.launcher.ray.cluster.setup_commands=[]",
+                "hydra.launcher.env_setup.commands=[]",
```
why do you need to override the commands?

jieru-hu replied:
:)
This is due to our test instance doesn't have access to pip so the setup commands would not work during integration testing.

omry replied:
Gotcha. add a comment about it?

jieru-hu replied:
thanks, added.

omry commented on plugins/hydra_ray_launcher/tests/test_ray_aws_launcher.py:
```diff
@@ -294,15 +294,15 @@ class TestRayAWSLauncher(LauncherTestSuite):
-                f"+hydra.launcher.ray_cluster_cfg.worker_nodes={ray_nodes_conf_override}",
-                f"+hydra.launcher.ray_cluster_cfg.head_node={ray_nodes_conf_override}",
+                f"hydra.launcher.ray.cluster.provider.key_pair.key_name=hydra_test_{cluster_name}",
+                "hydra.launcher.ray.cluster.setup_commands=[]",
+                "hydra.launcher.env_setup.commands=[]",
```
here too.

omry commented on website/docs/plugins/ray_launcher.md:
```diff
@@ -34,81 +34,75 @@ defaults:
+  cluster:
+    cluster_name: default
+    min_workers: 0
+    max_workers: 1
+    initial_workers: 0
+    autoscaling_mode: default
+    target_utilization_fraction: 0.8
+    idle_timeout_minutes: 5
+    docker:
+      image: ''
+      container_name: ''
+      pull_before_run: true
+      run_options: []
+    provider:
+      type: aws
+      region: us-west-2
+      availability_zone: us-west-2a,us-west-2b
+      cache_stopped_nodes: false
+      key_pair:
+        key_name: hydra
+    auth:
+      ssh_user: ubuntu
+    head_node:
+      InstanceType: m5.large
+      ImageId: ami-008d8ed4bd7dc2485
+    worker_nodes:
+      InstanceType: m5.large
+      ImageId: ami-008d8ed4bd7dc2485
+    file_mounts: {}
+    initialization_commands: []
+    setup_commands: []
+    head_setup_commands: []
+    worker_setup_commands: []
+    head_start_ray_commands:
+    - ray stop
+    - ulimit -n 65536; ray start --head --redis-port=6379 --object-manager-port=8076
+      --autoscaling-config=~/ray_bootstrap_config.yaml
+    worker_start_ray_commands:
+    - ray stop
+    - ulimit -n 65536; ray start --address=$RAY_HEAD_IP:6379 --object-manager-port=8076
```
much nicer.

omry reviewed (approved):
shipit!
